cpp/src/arrow/filesystem/azurefs_test.cc

```diff
@@ -889,6 +889,117 @@ class TestAzureFileSystem : public ::testing::Test {
+  }
+
+  void TestDeleteFileAtSubdirectory(bool create_empty_dir_marker_first) {
+    auto data = SetUpPreexistingData();
+
+    auto setup_dir_file0 = [this, create_empty_dir_marker_first, &data]() {
+      if (WithHierarchicalNamespace()) {
+        ASSERT_FALSE(create_empty_dir_marker_first);
+        auto adlfs_client =
+            datalake_service_client_->GetFileSystemClient(data.container_name);
+        CreateFile(adlfs_client, "dir/file0", PreexistingData::kLoremIpsum);
+      } else {
+        auto container_client = CreateContainer(data.container_name);
+        if (create_empty_dir_marker_first) {
+          CreateBlob(container_client, "dir/", "");
+        }
+        CreateBlob(container_client, "dir/file0", PreexistingData::kLoremIpsum);
+      }
+    };
+    setup_dir_file0();
+
+    // Trying to delete a non-existing file in an existing directory should fail
+    EXPECT_RAISES_WITH_MESSAGE_THAT(
+        IOError,
+        ::testing::HasSubstr("Path does not exist '" + data.Path("dir/nonexistent-file") +
+                             "'"),
+        fs()->DeleteFile(data.Path("dir/nonexistent-file")));
+    EXPECT_RAISES_WITH_MESSAGE_THAT(
+        IOError,
+        ::testing::HasSubstr("Not a regular file: '" +
+                             data.Path("dir/nonexistent-file/") + "'"),
+        fs()->DeleteFile(data.Path("dir/nonexistent-file/")));
+
+    // Trying to delete the directory with DeleteFile should fail
+    if (WithHierarchicalNamespace()) {
+      EXPECT_RAISES_WITH_MESSAGE_THAT(
+          IOError, ::testing::HasSubstr("Not a regular file: '" + data.Path("dir") + "'"),
+          fs()->DeleteFile(data.Path("dir")));
+    } else {
+      EXPECT_RAISES_WITH_MESSAGE_THAT(
+          IOError, ::testing::HasSubstr("Path does not exist '" + data.Path("dir") + "'"),
+          fs()->DeleteFile(data.Path("dir")));
+    }
+    EXPECT_RAISES_WITH_MESSAGE_THAT(
+        IOError, ::testing::HasSubstr("Not a regular file: '" + data.Path("dir/") + "'"),
+        fs()->DeleteFile(data.Path("dir/")));
+
+    arrow::fs::AssertFileInfo(fs(), data.Path("dir"), FileType::Directory);
+    arrow::fs::AssertFileInfo(fs(), data.Path("dir/"), FileType::Directory);
+    arrow::fs::AssertFileInfo(fs(), data.Path("dir/file0"), FileType::File);
+    ASSERT_OK(fs()->DeleteFile(data.Path("dir/file0")));
+    arrow::fs::AssertFileInfo(fs(), data.Path("dir"), FileType::Directory);
+    arrow::fs::AssertFileInfo(fs(), data.Path("dir/"), FileType::Directory);
+    arrow::fs::AssertFileInfo(fs(), data.Path("dir/file0"), FileType::NotFound);
+
+    // Recreating the file on the same path gurantees leases were properly released/broken
+    setup_dir_file0();
+
+    arrow::fs::AssertFileInfo(fs(), data.Path("dir/file0"), FileType::File);
+    EXPECT_RAISES_WITH_MESSAGE_THAT(
+        IOError,
+        ::testing::HasSubstr("Not a regular file: '" + data.Path("dir/file0/") + "'"),
+        fs()->DeleteFile(data.Path("dir/file0/")));
```

Comment:
How about swapping these assertions to check `DeleteFile("dir/file0")` while `dir/file0` exists?

```suggestion
    EXPECT_RAISES_WITH_MESSAGE_THAT(
        IOError,
        ::testing::HasSubstr("Not a regular file: '" + data.Path("dir/file0/") + "'"),
        fs()->DeleteFile(data.Path("dir/file0/")));
    arrow::fs::AssertFileInfo(fs(), data.Path("dir/file0"), FileType::File);
```

Reply:
Done.